packages/backend-defaults/src/entrypoints/urlReader/lib/ConfluenceUrlReader.ts

```diff
@@ -0,0 +1,266 @@
+    return {
+      host,
+      space,
+      pageId,
+      title: pageTitle,
+    };
+  }
+
+  async readTree(
+    url: string,
+    _options?: UrlReaderServiceReadTreeOptions,
+  ): Promise<UrlReaderServiceReadTreeResponse> {
+    const { pageId } = this.parseUrl(url);
+    const pages = await this.loadConfluencePage(pageId, '/docs');
+
+    const index = this.generateIndexPage(pages); // Generate index page for fetched pages
+    pages.push({ data: Readable.from(index), path: `docs/index.md` });
+
+    return this.deps.treeResponseFactory.fromReadableArray(pages);
+  }
+
+  search(
+    _url: string,
+    _options?: UrlReaderServiceSearchOptions,
+  ): Promise<UrlReaderServiceSearchResponse> {
+    throw new Error('Method not implemented.');
+  }
+
+  async loadConfluencePage(
+    pageId: string,
+    path: string,
+  ): Promise<FromReadableArrayOptions> {
+    const docItems: FromReadableArrayOptions = [];
+
+    try {
+      // fetch page
+      const { page, title, lastModifiedAt } = await this.fetchConfluencePage(
+        pageId,
+      );
+
+      // fetch its attachments
+      const attachments: Attachment[] =
+        await this.fetchConfluencePageAttachments(pageId);
+      for (const attachment of attachments) {
+        const imageData: any = await this.fetchAttachment(attachment);
+        docItems.push({
+          data: Readable.fromWeb(imageData),
+          path: `${path}/attachments/${attachment.title.replaceAll(' ', '-')}`,
+          lastModifiedAt: new Date(attachment.version.createdAt),
+        });
+      }
+
+      const pageMarkdown = this.replaceAttachmentLinks(attachments, page);
+      docItems.push({
+        data: Readable.from(pageMarkdown),
+        path: `${path}/${title}.md`,
+        lastModifiedAt: new Date(lastModifiedAt),
+      });
+
+      // fetch child pages and repeat
+      const children = await this.fetchChildPages(pageId);
+      for (const childPage of children) {
+        const childPageContent = await this.loadConfluencePage(
+          childPage.id,
+          `${path}/${title.toLocaleLowerCase().replaceAll(' ', '-')}`,
+        );
+        docItems.push(...childPageContent);
+      }
+    } catch (error) {
+      console.log(error);
+      this.logger.error(
+        `Error reading page ${pageId} from confluence: ${error.message}`,
+      );
+    }
+
+    return docItems;
+  }
+
+  private replaceAttachmentLinks(attachments: Attachment[], markdown: string) {
+    let markdownToPublish = markdown;
+    for (const attachment of attachments) {
+      const attachmentTitle = attachment.title.replace(/ /g, '-');
+      const attachmentTitleInUrl = attachment.title.replace(/ /g, '%20');
+      const regex = attachmentTitleInUrl.includes('.pdf')
+        ? new RegExp(`(\\[.*?\\]\\()(.*?${attachmentTitleInUrl}.*?)(\\))`, 'gi')
+        : new RegExp(
+            `(\\!\\[.*?\\]\\()(.*?${attachmentTitleInUrl}.*?)(\\))`,
+            'gi',
+          );
```

Comment:
This is a little dense to read, maybe break it out a bit or add some comments

Reply:
Added some comments to describe it. could you check if that's fine